src/hooks.rs

```diff
@@ -0,0 +1,45 @@
+//! This module provides a way to register decoding hooks for image formats not directly supported
+//! by this crate.
+
+use std::{
+    collections::HashMap,
+    io::{BufReader, Read, Seek},
+    sync::RwLock,
+};
+
+use crate::{ImageDecoder, ImageFormat, ImageResult};
+
+pub(crate) trait ReadSeek: Read + Seek {}
+impl<T: Read + Seek> ReadSeek for T {}
+
+pub(crate) static DECODING_HOOKS: RwLock<Option<HashMap<ImageFormat, DecodingHook>>> =
+    RwLock::new(None);
+
+/// A wrapper around a type-erased trait object that implements `Read` and `Seek`.
+pub struct BoxReadSeek<'a>(pub(crate) Box<dyn ReadSeek + 'a>);
+impl Read for BoxReadSeek<'_> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        self.0.read(buf)
+    }
+}
+impl Seek for BoxReadSeek<'_> {
+    fn seek(&mut self, pos: std::io::SeekFrom) -> std::io::Result<u64> {
+        self.0.seek(pos)
+    }
+}
+
+/// A function to produce an `ImageDecoder` for a given image format.
+pub type DecodingHook = Box<
+    dyn for<'a> Fn(BufReader<BoxReadSeek<'a>>) -> ImageResult<Box<dyn ImageDecoder + 'a>>
+        + Send
+        + Sync,
+>;
+
+/// Register a new decoding hook or replace an existing one.
+pub fn register_decoding_hook(format: ImageFormat, hook: DecodingHook) {
```

Comment:
Maybe replacing shouldn't be allowed?

The application controls `main`, so it can always be the first to register its preferred codecs. But it could be annoying if some dependency later "helpfully" replaced that with its own preferred codecs. 

Especially if there's no way to check if there's an existing hook already, crates that need a certain format supported won't have much choice except to always register.

Reply:
Switched this to blocking replacement. Also added a `decoding_hook_registered` to check whether a hook has been registered for a given format